### ISSUE TITLE
Attempting to validate a object without discriminator field results in exception

### ISSUE THREAD

**Issue by rafael-org** [byu-oit/openapi-enforcer#108]
When a schema have a discriminator property and this property is provided an invalid value, it results in an exception. It did not happen on version 1.10.0. I believe this started to happen somewhere along the 1.10 version, probably on 1.10.1 or 1.10.8.

Definition exemple:
```
      oneOf:
        - $ref: '#/components/schemas/content.text'
        - $ref: '#/components/schemas/content.file'
      discriminator:
        propertyName: type
        mapping:
          text: '#/components/schemas/content.text'
          file: '#/components/schemas/content.file'
```
          
If so instead of using `type` as either `text` or `file`, I use something like `audio` I no longer get a validation error and instead get a exception:
`Cannot read property 'type' of undefined TypeError: Cannot read property 'type' of undefined`

And happens at:
`at runValidate (/home/rafael.souza/git/api-control/node_modules/openapi-enforcer/src/schema/validate.js:82:86)`

It seems to fail on the `value[key]` (value is undefined for some reason in this case).

**Comment by Gi60s** [comment]
Hello @rafael-org! Are you running the latest version of the OpenAPI Enforcer and seeing this bug? I just created and ran the following tests without any problems, so if you're on the latest version of the enforcer then I'll need some more help recreating the scenario so I can properly test and fix it.

Here is the test I ran:

```js
const def = {
    openapi: '3.0.0',
    info: { title: '', version: '' },
    paths: {},
    components: {
        schemas: {
            'content.text': {
                type: 'object',
                properties: {
                    type: {
                        type: 'string'
                    },
                    content: {
                        type: 'string'
                    }
                }
            },
            'content.file': {
                type: 'object',
                properties: {
                    type: {
                        type: 'string'
                    },
                    content: {
                        type: 'string',
                        format: 'byte'
                    }
                }
            },
            content: {
                oneOf: [
                    { $ref: '#/components/schemas/content.text' },
                    { $ref: '#/components/schemas/content.file' }
                ],
                discriminator: {
                    propertyName: 'type',
                    mapping: {
                        text: '#/components/schemas/content.text',
                        file: '#/components/schemas/content.file'
                    }
                }
            }
        }
    }
}

it('should validate correctly with valid discriminator property name', async () => {
    const openapi = await Enforcer(def, { hideWarnings: true })
    const err = openapi.components.schemas.content.validate({
        type: 'text',
        content: 'hello'
    })
    expect(err).to.equal(undefined)
})

it('should produce a valid exception with invalid discriminator property name', async () => {
    const openapi = await Enforcer(def, { hideWarnings: true })
    const err = openapi.components.schemas.content.validate({
        type: 'audio',
        content: 'hello'
    })
    expect(err).to.match(/Discriminator property "type" as "audio" did not map to a schema/)
})
```

**Comment by rafael-org** [comment]
I tried many versions. On the actual API we were using 1.10.0 (I think because of this), I have been trying to update to 1.12.5 (on the 1.13.1 I stumbled on the https://github.com/byu-oit/openapi-enforcer/issues/109, thanks for the fix by the way, it worked).

This is the specification I am trying to test [openapi.yaml.txt](https://github.com/byu-oit/openapi-enforcer/files/5905531/openapi.yaml.txt). I get this error on a test with the following payload on `/v2/channels/whatsapp/messages`:

```
      {
        from: 'SENDER',
        to: 'RECIPIENT',
        contents: [
          {
            type: 'attachment',
          },
        ],
      }
```

I gonna try to test it using you code to see if it also happens.

**Comment by rafael-org** [comment]
We use `{ fullResult: true }`, I think that must be related to the issue.

On your code when I use it both tests fails because no `.components` is undefined.

**Comment by Gi60s** [comment]
Are you saying that if I change the tests to use the `fullResult` option that you're seeing that cause test failures?

Here is my update (minus the `def` value which I've left unchanged), but I'm not seeing any test failure:

```js
it('should validate correctly with valid discriminator property name', async () => {
    const [ openapi, error ] = await Enforcer(def, { fullResult: true })
    expect(error).to.equal(undefined)

    const err = openapi.components.schemas.content.validate({
        type: 'text',
        content: 'hello'
    })
    expect(err).to.equal(undefined)
})

it('should produce a valid exception with invalid discriminator property name', async () => {
    const [ openapi, error ] = await Enforcer(def, { fullResult: true })
    expect(error).to.equal(undefined)
    
    const err = openapi.components.schemas.content.validate({
        type: 'audio',
        content: 'hello'
    })
    expect(err).to.match(/Discriminator property "type" as "audio" did not map to a schema/)
})
```

**Comment by rafael-org** [comment]
Indeed I was not destructuring it. I am going to test with our spec to see if I can pinpoint the exactly issue I am having.

**Comment by rafael-org** [comment]
Okay, two things I noticed testing:
* Just by creating a simple specification I was unable to replicate the issue
* Validating using the schema the exception do not occur: it only happens doing request validations
* Also, it only happens when the object have properties that do not match any schema.
* Testing with `production: true` and `production: false` produced a tiny bit different results

Debugging I noticed that when the error occurs, the `value` variable that suppose to be `{ from: 'SENDER', to: 'RECEIVER', contents: [ [Object] ] }` was `{ from: 'SENDER', to: 'RECEIVER', contents: [ undefined ] }` (and with `production: true` it was `{ from: 'SENDER', to: 'RECEIVER', contents: [ ] }`).
* So, for some reason the sequence of events to get a request validate against our spec was resulting in the "original" (not the actual original) contents array being changed.
* Digging a little deeper I found it seemed to be related to the deserialization of the body

But now I feel like I am chasing around my tail. But I did found that the array gets changed somewhere in the call of the `schema.deserialize` on https://github.com/byu-oit/openapi-enforcer/blob/master/src/enforcers/Operation.js#L651 (the object is okay before the call and not okay afterwards). Because of this, later it fails to generate the proper error message on the following stack:
```
  https://github.com/byu-oit/openapi-enforcer/issues/108
    Zenvia spec
      schema message.whatsapp
        ✓ should validate correctly (611ms)
        ✓ should produce a valid exception with invalid schema (377ms)
      post on messages
        ✓ should validate correctly (419ms)
        ✓ should produce a valid exception with invalid schema (text property) (383ms)
        1) should produce a valid exception with invalid schema (url property)


  4 passing (2s)
  1 failing

  1) https://github.com/byu-oit/openapi-enforcer/issues/108
       Zenvia spec
         post on messages
           should produce a valid exception with invalid schema (url property):
     TypeError: Cannot read property 'type' of undefined
      at runValidate (node_modules/openapi-enforcer/src/schema/validate.js:82:86)                                                                                                                                                                                              
      at value.forEach (node_modules/openapi-enforcer/src/schema/validate.js:136:21)                                                                                                                                                                                           
      at Array.forEach (<anonymous>)                                                                                                                                                                                                                                           
      at runValidate (node_modules/openapi-enforcer/src/schema/validate.js:135:23)                                                                                                                                                                                             
      at keys.forEach.key (node_modules/openapi-enforcer/src/schema/validate.js:168:21)                                                                                                                                                                                        
      at Array.forEach (<anonymous>)                                                                                                                                                                                                                                           
      at runValidate (node_modules/openapi-enforcer/src/schema/validate.js:160:18)                                                                                                                                                                                             
      at schema.allOf.forEach (node_modules/openapi-enforcer/src/schema/validate.js:48:13)                                                                                                                                                                                     
      at Array.forEach (<anonymous>)                                                                                                                                                                                                                                           
      at runValidate (node_modules/openapi-enforcer/src/schema/validate.js:47:22)                                                                                                                                                                                              
      at Schema.validate (node_modules/openapi-enforcer/src/enforcers/Schema.js:215:9)                                                                                                                                                                                         
      at deserializeAndValidate (node_modules/openapi-enforcer/src/enforcers/Operation.js:652:42)                                                                                                                                                                              
      at Operation.request (node_modules/openapi-enforcer/src/enforcers/Operation.js:294:33)                                                                                                                                                                                   
      at OpenApi.request (node_modules/openapi-enforcer/src/enforcers/OpenApi.js:146:38)                                                                                                                                                                                       
      at Context.it (test/routes/openapi-enforcer-issue-108.spec.ts:111:33)
```

I am starting to guess that it might be because we also are using allOf, both on root element as on the content array.

Here is the test I adapted from yours using our spec/def (it is in a [yaml file](https://github.com/byu-oit/openapi-enforcer/files/5910502/openapi.yaml.txt)). Only the last (fifth) test fails.

I will later try to chase it a bit more.

```
import * as Enforcer from 'openapi-enforcer';
import { expect } from 'chai';
import * as yaml from 'js-yaml';
import * as fs from 'fs';
import * as path from 'path';

describe.only('https://github.com/byu-oit/openapi-enforcer/issues/108', () => {

  describe('Zenvia spec', () => {

    const def = yaml.safeLoad(
      fs.readFileSync(path.join('config', 'openapi.yaml'), 'utf-8'),
    );

    describe('schema message.whatsapp', () => {

      it('should validate correctly', async () => {
        const [openapi, error] = await Enforcer(def, { fullResult: true, componentOptions: {} });
        expect(error).to.equal(undefined);

        const err = openapi.components.schemas['message.whatsapp'].validate({
          from: 'SENDER',
          to: 'RECEIVER',
          contents: [{
            type: 'text',
            text: 'hello',
          }],
        });

        expect(err).to.equal(undefined);
      });

      it('should produce a valid exception with invalid schema', async () => {
        const [openapi, error] = await Enforcer(def, { fullResult: true, componentOptions: {} });
        expect(error).to.equal(undefined);

        const err = openapi.components.schemas['message.whatsapp'].validate({
          from: 'SENDER',
          to: 'RECEIVER',
          contents: [{
            type: 'audio',
            content: 'hello',
          }],
        }, { readWriteMode: 'write' });

        expect(err).to.match(/Discriminator property "type" as "audio" did not map to a schema/);
      });

    });

    describe('post on messages', () => {

      it('should validate correctly', async () => {
        const [openapi, error] = await Enforcer(def, { fullResult: true, toString: true });
        expect(error).to.equal(undefined);

        const [, err] = openapi.request({
          path: '/channels/whatsapp/messages',
          method: 'POST',
          query: {
          },
          headers: {
            'content-type': 'application/json',
            'x-api-token': 'some-api-token',
          },
          body: {
            from: 'SENDER',
            to: 'RECEIVER',
            contents: [{
              type: 'text',
              text: 'hello',
            }],
          },
        }, { allowOtherQueryParameters: [] });

        expect(err).to.equal(undefined);
      });

      it('should produce a valid exception with invalid schema (text property)', async () => {
        const [openapi, error] = await Enforcer(def, { fullResult: true, toString: true, componentOptions: { production: false } });
        expect(error).to.equal(undefined);

        const [, err] = openapi.request({
          path: '/channels/whatsapp/messages',
          method: 'POST',
          query: {
          },
          headers: {
            'content-type': 'application/json',
            'x-api-token': 'some-api-token',
          },
          body: {
            from: 'SENDER',
            to: 'RECEIVER',
            contents: [{
              type: 'audio',
              text: 'hello',
            }],
          },
        }, { allowOtherQueryParameters: [] });

        expect(err.toString()).to.match(/Discriminator property "type" as "audio" did not map to a schema/);

      });

      it('should produce a valid exception with invalid schema (url property)', async () => {
        const [openapi, error] = await Enforcer(def, { fullResult: true, toString: true, componentOptions: { production: false } });
        expect(error).to.equal(undefined);

        const [, err] = openapi.request({
          path: '/channels/whatsapp/messages',
          method: 'POST',
          query: {
          },
          headers: {
            'content-type': 'application/json',
            'x-api-token': 'some-api-token',
          },
          body: {
            from: 'SENDER',
            to: 'RECEIVER',
            contents: [{
              type: 'audio',
              url: 'hello',
            }],
          },
        }, { allowOtherQueryParameters: [] });

        expect(err.toString()).to.match(/Discriminator property "type" as "audio" did not map to a schema/);

      });

    });

  });

});
```

But, even though I did not find the source yet, perhaps you can with this information.

**Comment by Gi60s** [comment]
Thank you for all the research and for providing a working example of the bug. As I dug into this I found a flaw in the deserialize logic when it comes to discriminators. I also decided to look at related functions such as serialize, populate, and random and found similar logic gaps. Ironically the `validate` was the only function that did not have the bug and that's why my initial posts on this thread did not show the problem.

I've published a fix for this to NPM as version `1.13.3`. Thank you again for the help.
